rfcs/000-no-std-support.md

```diff
@@ -0,0 +1,996 @@
+
+### `LlClientConnection`
+
+The lower-level `ClientConnection` API (to be) proposed as a solution to [issue #1362][gh1362] does not rely on the `io::{Read, Write}` traits.
+This new API -- from now on, referred to as `LlClientConnection` -- instead of the existing one (as of v0.21.6), will be exposed in no-std mode.
+
+[gh1362]: https://github.com/rustls/rustls/issues/1362
+
+### `SystemTime`
+
+`std::time::SystemTime` has an OS specific representation (e.g. UNIX-like systems use `timespec`, Windows uses `FILETIME`).
+An alternative system time representation that is OS-agnostic is a UNIX timestamp: e.g. seconds since `UNIX_EPOCH`.
+`SystemTime` can already be easily converted into a UNIX timestamp.
+
+A simple `UnixTimestamp` data type with no methods will be added to rustls.
+
+```rust
+/// `duration` since UNIX_EPOCH
+pub struct UnixTimestamp(pub core::time::Duration);
+```
+
+When the "std" feature is enabled, `TryFrom` implementations will be provided.
+
+```rust
+#[cfg(feature = "std")]
+mod std_impls {
+    use crate::UnixTimestamp;
+
+    use std::time::{SystemTime, SystemTimeError, UNIX_EPOCH};
+
+    impl TryFrom<SystemTime> for UnixTimestamp {
+        type Error = SystemTimeError;
+
+        fn try_from(st: SystemTime) -> Result<Self, Self::Error> {
+            Ok(Self(st.duration_since(UNIX_EPOCH)?))
+        }
+    }
+
+    impl TryFrom<UnixTimestamp> for SystemTime {
+        type Error = ();
+
+        fn try_from(ts: UnixTimestamp) -> Result<Self, Self::Error> {
+            UNIX_EPOCH.checked_add(ts.0).ok_or(())
+        }
+    }
+}
+```
+
+All public API will change to use `rustls::UnixTimestamp` instead of `std::time::SystemTime`.
+This is a **breaking change**.
+
+```rust
+pub trait ServerCertVerifier: Send + Sync {
+    fn verify_server_cert(
+        &self,
+        end_entity: &Certificate,
+        intermediates: &[Certificate],
+        server_name: &ServerName,
+        ocsp_response: &[u8],
+        now: UnixTimestamp, // <-
+    ) -> Result<ServerCertVerified, Error>;
+}
+
+pub trait ClientCertVerifier: Send + Sync {
+    fn verify_client_cert(
+        &self,
+        end_entity: &Certificate,
+        intermediates: &[Certificate],
+        now: UnixTimestamp, // <-
+    ) -> Result<ClientCertVerified, Error>;
+}
+
+pub fn verify_server_cert_signed_by_trust_anchor(
+    cert: &ParsedCertificate,
+    roots: &RootCertStore,
+    intermediates: &[Certificate],
+    now: UnixTimestamp, // <-
+) -> Result<(), Error> { /* .. */ }
+```
+
+Note that this also includes `ClientCertVerifier` even though it won't be exposed in no-std mode during the first stage.
+This is so a similar breaking change does not have to happen in the second stage where server functionality is made no-std compatible.
+
+### `SystemTime::now`
+
+Additionally, the client internally calls `SystemTime::now` prior to invoking `verify_server_cert`.
+In no-std mode, `ClientConfig` will gain a public `TimeProvider` setting.
```

Comment:
Though that leaves a bit of a hole for the `Ring` `CryptoProvider` in `std` mode.

Reply:
For time, I think we should try to wire up `webpki::Time` directly into rustls and just use that pervasively, perhaps via (something like) rustls-pki-types?

(Not sure the change to move time acquisition into `CryptoProvider` makes sense to me. Yes, randomness and time are both services provided by the OS, but time is otherwise not at all pertinent to cryptographic operations and seems much more aligned with the operating system/platform on which you're deploying.)

Reply:
> For time, I think we should try to wire up webpki::Time directly into rustls

there's also [`TimeBase`](https://github.com/rustls/rustls/blob/v/0.21.6/rustls/src/ticketer.rs#L15) which is used by the ticketer component and it's pretty much the same as `UnixTimestamp` proposed here.

The reason I didn't suggest making `TimeBase` public is that a `TimeProvider` will need to directly construct a `UnixTimestamp` and `TimeBase`'s field is private and lacks any `fn(Duration) -> Self` constructor.

> without leaking a webpki type into the rustls public API) and just use that pervasively

then it wouldn't be suitable for use by `TimeProvider` as that requires a public type with a direct constructor or a public field.

Reply:
I'm just saying it would be nice to standardize all of `webpki::Time`, `rustls::TimeBase` and whatever we're using `SystemTime` for today and roll it into a single type, because it seems all of these are trying to do roughly the same thing anyway.

Reply:
from looking at platform-verifiers
- [android](https://github.com/rustls/rustls-platform-verifier/blob/19d530466154f5a531f607cb94b71c8006a3d02a/src/verification/android.rs#L86-L91) uses a unix timestamp in millis
- [macos](https://github.com/rustls/rustls-platform-verifier/blob/19d530466154f5a531f607cb94b71c8006a3d02a/src/verification/apple.rs#L19-L38) computes a unix timestamp `Duration` and then maps it into its own CFDate type
- [windows](https://github.com/rustls/rustls-platform-verifier/blob/main/src/verification/windows.rs#L346-L365) does roughly the same as macos but uses its own FILETIME type

I think whether to represent the unix timestamp as `u64` seconds (`webpki::Time`) or `core::time::Duration` (like `TimeBase` does) depends on whether the protocol needs sub-second resolution or not. given that Android uses millisecond resolution probably it does?

Reply:
In the context of certificate validation sub-second resolution seems like overkill :thinking: 

I don't feel strongly but if this is the only point of contention for this draft RFC it might be nice to figure out what we need to do to reach consensus.

Reply:
https://github.com/rustls/pki-types/pull/7